src/wh_server_crypto.c

```diff
@@ -0,0 +1,339 @@
+        if (foundIndex == -1 && (server->cache[i].meta->id == 0 ||
+            server->cache[i].meta->id == meta->id)) {
+            foundIndex = i;
+        }
+    }
+#if 0
+    /* if no empty slots, check for a commited key we can evict */
+    if (foundIndex == -1) {
+        for (i = 0; i < WOLFHSM_CACHE_COUNT; i++) {
+            if (ctx->cache[i].commited == 1) {
+                foundIndex = i;
+                break;
+            }
+        }
+    }
+    /* return error if we are out of cache slots */
+    if (foundIndex == -1)
+        return MEMORY_E;
+    ctx->cache[foundIndex].commited = 0;
+    /* check if the key is in nvm, mark as commited */
+    for (i = 0; i < WOLFHSM_KEYSLOT_COUNT; i++) {
+        /* check for duplicate id, shouldn't cache a key in nvm */
+        if (ctx->nvmMetaCache[i].id == meta->id)
+            ctx->cache[foundIndex].commited = 1;
+    }
+#endif
+    /* write key if slot found */
+    XMEMCPY((uint8_t*)server->cache[foundIndex].buffer, in, meta->len);
+    XMEMCPY((uint8_t*)server->cache[foundIndex].meta, (uint8_t*)meta,
+        sizeof(whNvmMetadata));
+#ifdef WOLFHSM_SHE_EXTENSION
+    /* if this was RAM_KEY, set the global so we can export */
+    if (meta->id == WOLFHSM_SHE_TRANSLATE_KEY_ID(WOLFHSM_SHE_RAM_KEY_ID))
+        hsmSheRamKeyPlain = 1;
+#endif
+    return 0;
+}
+
+static int hsmReadKey(whServerContext* server, whNvmMetadata* meta, uint8_t* out)
+{
+    int i;
+    uint16_t searchId = meta->id;
+    uint16_t outLen = meta->len;
+    /* make sure id is valid */
+    if (meta->id == 0)
+        return BAD_FUNC_ARG;
+    /* check the cache */
+    for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+        /* copy the meta and key before returning */
+        if (server->cache[i].meta->id == searchId) {
+            /* check outLen */
+            if (server->cache[i].meta->len > outLen)
+                return BUFFER_E;
+            XMEMCPY((uint8_t*)meta, (uint8_t*)server->cache[i].meta,
+                sizeof(whNvmMetadata));
+            XMEMCPY(out, server->cache[i].buffer, meta->len);
+            return 0;
+        }
+    }
+    return BAD_FUNC_ARG;
+#if 0
+    /* setup address */
+    if (ctx->partition == 1)
+        addr = WOLFHSM_PART_1;
+    /* find id in list */
+    for (i = 0; i < WOLFHSM_KEYSLOT_COUNT; i++) {
+        /* check id */
+        if (ctx->nvmMetaCache[i].id == searchId)
+            break;
+        /* keep track of length used by previous keys */
+        keyOffset += ctx->nvmMetaCache[i].len;
+    }
+    /* if we looped to the end, no match */
+    if (i >= WOLFHSM_KEYSLOT_COUNT)
+        ret = BAD_FUNC_ARG;
+    /* check the length, meta->len is the return size */
+    if (ret == 0 && ctx->nvmMetaCache[i].len > outLen)
+        ret = BUFFER_E;
+    /* read key */
+    if (ret == 0) {
+        ret = hal_flash_read(addr + WOLFHSM_HEADER_SIZE + keyOffset, out,
+            ctx->nvmMetaCache[i].len);
+    }
+    /* calculate key digest */
```

Comment:
Recommend to make integrity/auth of keys to be optional and not required at this point.  No changes for now.